### ISSUE TITLE
unstable_cache does not deserialize dates

### ISSUE THREAD

**Issue by SSardorf** [vercel/next.js#51613]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Mon Apr 24 20:52:24 PDT 2023; root:xnu-8796.121.2~5/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.12.1
      npm: 9.6.4
      Yarn: 1.22.19
      pnpm: 7.13.1
    Relevant packages:
      next: 13.4.7-canary.2
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 4.9.5
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true), Data fetching (gS(S)P, getInitialProps)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/SSardorf/next_unstable_cache_date_deserialization

### To Reproduce

1.  `npm i`
2. `npx prisma generate`
3.  `npx prisma db push`
4. `npx prisma db seed`
5. `npm run dev`

When you first run the project, you'll see that it loads correctly (cause first prisma invocation will result in non-cached data)
When you refresh, the cached data will be shown, and it'll throw an error, as `post.createdAt` is no longer a Date, but a string:
`<p>{post.createdAt.toISOString()}</p>`

### Describe the Bug

When you first run the project, you'll see that it loads correctly (cause first prisma invocation will result in non-cached data)
When you refresh, the cached data will be shown, and it'll throw an error, as `post.createdAt` is no longer a Date, but a string:
`<p>{post.createdAt.toISOString()}</p>`

### Expected Behavior

unstable_cache should correctly de-serialize Dates.

**Comment by SSardorf** [comment]
I see it's also set as a TODO in the unstable_cache function
https://github.com/vercel/next.js/blob/36a7aff6d656f81a62a66d5f9dfd3b173a7c26d9/packages/next/src/server/web/spec-extension/unstable-cache.ts#L82-L87

**Comment by numman-ali** [comment]
I came across this problem today and found a solution that keeps everything intact

```
import { AppClientEnum, type ChallengeDtoModel } from "@getcub3/api-models";
import { challengeService } from "@getcub3/api-services";
import { unstable_cache as cache } from "next/cache";
import { parse, stringify } from "superjson";

import { dbConnect } from "@/lib";

export const getChallengeById = async (
  id: string,
): Promise<ChallengeDtoModel | null> => {
  const cachedResult: string | null = await cache(
    async () => {
      await dbConnect();
      const challenge = await challengeService.getChallengeById(
        id,
        AppClientEnum.Public,
      );
      return stringify({ ...challenge });
    },
    [`public-challenge-${id}`],
    {
      tags: [`public-challenge-${id}`],
      revalidate: 60,
    },
  )();
  if (!cachedResult) {
    return null;
  }

  return parse<ChallengeDtoModel>(cachedResult);
};
```

1. `{ ...challenge}` will convert class to plain object
2. `stringify` from superjson will make a SuperJSON string that keeps values intact
3. `parse` from superjson will parse the SuperJSON string into a plain object that has the correct types 

Doing it this way means I can still use class-transformer class as a single source of truth for my types as well

**Comment by andrew-gropyus** [comment]
From some cursory looks at the source of react/cache and nextjs/unstable_cache, it seems that react/cache does not do any serialisation of objects however nextjs/unstable_cache does.

If it is intended that nextjs/unstable_cache to ultimately be a drop in replacement for react/cache then this might cause some issues.


In addition to SSardorf's example above where they are caching database responses, I would add another example of caching grpc responses, which can also have Date objects in them.

**Comment by mordechaim** [comment]
Based on the suggestion from @numman-ali, here's a utility you can use:

```js
import { unstable_cache } from 'next/cache'
import superjson from 'superjson'

export const cache = (fn, keys, opts) => {
    const wrap = async (...params) => {
        const result = await fn(...params)
        return superjson.stringify(result)
    }

    const cachedFn = unstable_cache(wrap, keys, opts)

    return async (...params) => {
        const result = await cachedFn(...params)
        return superjson.parse(result)
    }
}
```

**Comment by chrishoermann** [comment]
here's a typed version from @mordechaim's solution:

```ts
import { unstable_cache } from "next/cache";
import { parse, stringify } from "superjson";

export const cache = <T, P extends unknown[]>(
  fn: (...params: P) => Promise<T>,
  keys: Parameters<typeof unstable_cache>[1],
  opts: Parameters<typeof unstable_cache>[2],
) => {
  const wrap = async (params: unknown[]): Promise<string> => {
    const result = await fn(...(params as P));
    return stringify(result);
  };

  const cachedFn = unstable_cache(wrap, keys, opts);

  return async (...params: P): Promise<T> => {
    const result = await cachedFn(params);
    return parse(result);
  };
};
```

**Comment by JDuqn** [comment]
Seconding this.

I am currently using the solution given by @mordechaim and @chrishoermann 

Something built-in would be nice.

**Comment by kachkaev** [comment]
Looks like the same applies to Set():

```ts
const getCachedSet = unstable_cache(async () => {
  return new Set(["hello"]);
}, ["cached-set"]);
```

```ts
console.log((await getCachedSet()).has("hello"));
```

↑ This sometimes logs `true` and sometimes throws:

```
⨯ TypeError: (intermediate value).has is not a function
   at CatalogContents (./app/page.tsx:42:42)
```

Adding `console.log(await getCachedSet())` returns different values between page re-renders:

```
Set(1) { 'hello' }
{} object
{} object
```

**Comment by mabasic** [comment]
I would like to point out that not only that it returns Date as string, it reports Typescript type as unchanged. Thus making the entire codebase using that data invalid since TS reports Date, but the data is actually string.

**Comment by controversial** [comment]
I agree, at the very least the Typescript types need to be updated to reflect that the returned value will be `string` instead of `Date`. The existing function type is dangerous and leads to runtime errors

**Comment by mordechaim** [comment]
Or just use [this](https://github.com/vercel/next.js/issues/51613#issuecomment-1892644565) and call it a day...

**Comment by controversial** [comment]
yep, just offering improved type declarations for `unstable_cache` as an alternative measure to improve type safety of the *built-in* serialization behavior and avoid unexpected runtime errors without having to add an external library

**Comment by mordechaim** [comment]
If I understand correctly, nextjs already uses superjson library internally to serialize RSC props passed to client components. So not necessarily an external library

**Comment by mordechaim** [comment]
The new [`"use cache"`](https://nextjs.org/docs/canary/app/api-reference/directives/use-cache) directive will now be able to cache any RSC payload, including dates and many more.

**Comment by GustavoOS** [comment]
This issue applies to other formats like bigint as well. Also fixable using superjson.

**Comment by chungweileong94** [comment]
It's okay that this stays as it is, but this needs to be documented tho. I know Next.js can serialize certain objects over the wire, but I didn't expect this.

**Comment by dantheli** [comment]
**Be careful when using the custom `cache` solution mentioned above!** `unstable_cache` by default uses the passed-in function's definition and its arguments as the cache key. If this wrapper solution is used, then the function definition of `wrap`, which does not change, and any arguments passed to `fn` will be part of the cache key. Therefore, if you plan to use the `cache` solution for multiple cache entries, unique `keyParts` MUST be passed in, in order to differentiate the cache entries.

I'm embarrassed to say that I scratched my head for two days about why my cache was returning the wrong data, only to read into what `unstable_cache` was doing.

Please see [this discussion thread](https://github.com/vercel/next.js/discussions/63099) for more info.

We can maybe improve the `cache` solution by always including `fn.toString()` as a key part (somebody check my work as I'm not a typescript person), which is what `unstable_cache` would do.

```ts
import { unstable_cache } from 'next/cache'
import { parse, stringify } from 'superjson'

type Callback = (...args: any[]) => Promise<any>

export function cache<T extends Callback>(
  fn: T,
  keyParts?: string[],
  opts?: { revalidate?: number | false; tags?: string[] }
): T {
  const wrap = async (params: Parameters<T>): Promise<string> => {
    const result = await fn(...params)
    return stringify(result)
  }

  return (async (...params: Parameters<T>) => {
    const keyPartsWithFn = [
      fn.toString(), // Use the function itself as part of the cache key
      ...(keyParts ?? []),
    ]
    const result = await unstable_cache(wrap, keyPartsWithFn, opts)(params)
    return parse(result)
  }) as T
}
```

**Comment by mobeigi** [comment]
> **Be careful when using the custom `cache` solution mentioned above!** `unstable_cache` by default uses the passed-in function's definition and its arguments as the cache key. If this wrapper solution is used, then the function definition of `wrap`, which does not change, and any arguments passed to `fn` will be part of the cache key. Therefore, if you plan to use the `cache` solution for multiple cache entries, unique `keyParts` MUST be passed in, in order to differentiate the cache entries.
> 
> I'm embarrassed to say that I scratched my head for two days about why my cache was returning the wrong data, only to read into what `unstable_cache` was doing.
> 
> Please see [this discussion thread](https://github.com/vercel/next.js/discussions/63099) for more info.
> 

Thanks for this post @dantheli. On my website I have been having a weird cache inconsistency issue that has been maddening to fix. I feel like what you mention might be the reason why. I will try out the adjustment to see if it fixes my issue.

Next.js should just address the issue at it's its root. Manual serialization and deserialization is tedious and using hacky wrappers is also not very clean & maintainable.

**Comment by mordechaim** [comment]
The new [`"use cache"`](https://nextjs.org/docs/app/api-reference/directives/use-cache) directive handles dates perfectly.

It is the official stable replacement for `unstable_cache`

**Comment by mobeigi** [comment]
> The new [`"use cache"`](https://nextjs.org/docs/app/api-reference/directives/use-cache) directive handles dates perfectly.
> 
> It is the official stable replacement for `unstable_cache`

Thanks! I'll look into `use cache`. Is it functionally equivalent to `unstable_cache`?
Was migrating between `unstable_cache` usages and `use cache` easy enough?

**Comment by mordechaim** [comment]
Yes, pretty straightforward.

Chache tags are done via `cacheTag()` function, which allows you tag tag over the actual data! I don't know if you realize how big of a deal this is.

You can do:

```ts
const getBlog = async (id: number) => {
    "use cache"

    const blog = await fetchMyBlog(id)
    cacheTag('blogs', `blog:${id}`, `blog:${blog.slug}`)

    return blog
}
```

`revalidateTag` and `revalidatePath` continue to work as expected.

**Comment by mordechaim** [comment]
You can use the `experimental.useCache` flag which doesn't require enabling `dynamicIO`

**Comment by Jornve** [comment]
> Yes, pretty straightforward.
> 
> Cache tags are done via `cacheTag()` function, which allows you to tag over the actual data! I don't know if you realize how big of a deal this is.
> 
> You can do:
> 
> const getBlog = async (id: number) => {
>     "use cache"
> 
>     const blog = await fetchMyBlog(id)
>     cacheTag('blogs', `blog:${id}`, `blog:${blog.slug}`)
> 
>     return blog
> }
> `revalidateTag` and `revalidatePath` continue to work as expected.

Had the same problem with dates, using the "use cache" functionality resolved the issue.

**Comment by mobeigi** [comment]
I made the change today to transition from **unstable_cache_safe** to **use cache**.
Was relatively painless and seems to be working well with dates too.
Hopefully it also fixes my other cache inconsistency issue that was happening to me.
